storage/innobase/handler/handler0alter.cc

```diff
@@ -654,9 +654,19 @@ static MY_ATTRIBUTE((warn_unused_result)) bool innobase_need_rebuild(
   Alter_inplace_info::HA_ALTER_FLAGS alter_inplace_flags =
       ha_alter_info->handler_flags & ~(INNOBASE_INPLACE_IGNORE);
 
-  if (Encryption::none_explicitly_specified(
-          ha_alter_info->create_info->explicit_encryption,
-          ha_alter_info->create_info->encrypt_type.str) ||
+  const bool was_none_explicitly_specified{
+      Encryption::none_explicitly_specified(old_table->s->explicit_encryption,
+                                            old_table->s->encrypt_type.str)};
+
+  const bool is_none_explicitly_specified{Encryption::none_explicitly_specified(
+      ha_alter_info->create_info->explicit_encryption,
+      ha_alter_info->create_info->encrypt_type.str)};
+
+  const bool is_file_per_table{
+      !tablespace_is_shared_space(ha_alter_info->create_info)};
+
+  if ((!was_none_explicitly_specified && is_none_explicitly_specified &&
```

Comment:
Why do you need the was_none check?

Reply:
question 2: is there similar problem with ENCRYPTION=KEYRING? The OR part doesn't check the file_per_table?

Reply:
q1. Apparently when I do a rebuild without any change it can also cause a crash (discovered it when testing partition table) - so I make sure rebuild is needed.
q2. I do not see a problem with ENCRYPTION = KEYRING, apart from the fact that table with ENCRYPTION='KEYRING' can be created in MK encrypted tablespace. It is possible that Align KEYRING and MK encryption PR is already fixing this. I will create a bug for it, so it would not be forgotten.